[WIP] pq: add logbroker federation support (alternative 1, other is #16293)

yumkam opened the pull request:
### Changelog entry <!-- a user-readable short description of the changes that goes to CHANGELOG.md and Release Notes -->

...

### Changelog category <!-- remove all except one -->

* Not for changelog (changelog entry is not required)

### Description for reviewers <!-- (optional) description for those who read this PR -->

~~Change to YDB SDK (first two or three commits) will be applied separately.~~
Discover all topics in reader, pass partition list (up to maximum), per-cluster { name, endpoint, database, partition count } to row_dispatcher
Requires all clusters to be described at start of query

yumkam committed:
ydb-sdk: add TFederatedTopic::GetAllClusterInfo

asynchronous method to discover all clusters for federated topic

yumkam committed:
ydb-sdk: middle-level methods to demonstrate use of GetAllClusterInfo

may be omitted

yumkam committed:
ydb-sdk: IOStream << some more structures

yumkam committed:
pq: Add federated topic gateway methods

yumkam committed:
pq: support federated logbroker

uzhastik commented on ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic_impl.cpp:
```diff
@@ -49,6 +49,59 @@ void TFederatedTopicClient::TImpl::InitObserver() {
+                    dbinfo.Endpoint = db->endpoint();
+                    dbinfo.Path = db->path();
+                }
+                return std::move(result);
```
std::move избыточен?

uzhastik commented on ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic_impl.cpp:
```diff
@@ -49,6 +49,59 @@ void TFederatedTopicClient::TImpl::InitObserver() {
+                result.reserve(state->DbInfos.size());
+                for (const auto& db: state->DbInfos) {
+                    auto& dbinfo = result.emplace_back();
+                    switch(db->status()) {
```
нужен пробел после switch

uzhastik commented on ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic_impl.cpp:
```diff
@@ -49,6 +49,59 @@ void TFederatedTopicClient::TImpl::InitObserver() {
+NThreading::TFuture<std::vector<TFederatedTopicClient::TClusterInfo>> TFederatedTopicClient::TImpl::GetAllClusterInfo() {
+    InitObserver();
+    return Observer->WaitForFirstState().Apply(
+            [observer = Observer] (const auto &) {
```
цикл из умных указателей? предлагаю разорвать

yumkam replied:
Заменил на `std::weak_ptr`;
cons: кидает исключение в фьючу: если `FederatedTopicClient` уничтожен на момент срабатывания фьючи

uzhastik commented on ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic.cpp:
```diff
@@ -86,4 +86,67 @@ void TFederatedTopicClient::OverrideCodec(NTopic::ECodec codecId, std::unique_pt
+}
+
+void TFederatedTopicClient::TClusterInfo::AdjustTopicPath(std::string& path) const {
+    if (!Name.empty()) {
```
```
if (Name.empty()) {
  return;
}
```

uzhastik commented on ydb/public/sdk/cpp/include/ydb-cpp-sdk/client/federated_topic/federated_topic.h:
```diff
@@ -516,6 +518,33 @@ class TFederatedTopicClient {
+
+    static std::vector<NTopic::TTopicClient> GetAllTopicClients(const TDriver& driver, const std::vector<TClusterInfo>& clusterInfos, NTopic::TTopicClientSettings& clientSettings);
+
+    static std::vector<TAsyncDescribeTopicResult> DescribeAllTopics(const std::string& path, std::vector<NTopic::TTopicClient> &topicClients, const std::vector<TClusterInfo>& clusterInfos, NTopic::TDescribeTopicSettings& describeSettings);
```
амперсанды прижимаем влево

uzhastik commented on ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic.cpp:
```diff
@@ -86,4 +86,67 @@ void TFederatedTopicClient::OverrideCodec(NTopic::ECodec codecId, std::unique_pt
+            continue;
+        }
+        std::string adjustedPath = path;
+        info.AdjustTopicPath(adjustedPath);
```
adjustedPath надо использовать ниже?

uzhastik commented on ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic.cpp:
```diff
@@ -86,4 +86,67 @@ void TFederatedTopicClient::OverrideCodec(NTopic::ECodec codecId, std::unique_pt
+    results.reserve(topicClients.size());
+    for (size_t i = 0; i < topicClients.size(); ++i) {
+        const auto& info = clusterInfos[i];
+        if (!(info.Status == TClusterInfo::EStatus::AVAILABLE || info.Status == TClusterInfo::EStatus::READ_ONLY)) {
```
проверка дублируется в двух функциях. можно вынести в отдельную функцию

yumkam committed:
fix rebase errors

uzhastik commented on ydb/tests/fq/pq_async_io/mock_pq_gateway.cpp:
```diff
@@ -106,6 +111,21 @@ class TMockPqGateway : public IMockPqGateway {
     };
+
+    struct TMockFederatedTopicClient : public IFederatedTopicClient {
+        TMockFederatedTopicClient() {}
```
конструктор и деструктор можно убрать? кода меньше будет

uzhastik commented on ydb/core/fq/libs/row_dispatcher/topic_session.cpp:
```diff
@@ -556,6 +560,24 @@ void TTopicSession::CloseTopicSession() {
 
 void TTopicSession::TTopicEventProcessor::operator()(NYdb::NTopic::TReadSessionEvent::TDataReceivedEvent& event) {
+    auto minTime = Self.GetMinStartingMessageTimestamp();
+    auto &messages = event.GetMessages();
```
амперсанды прижимаем влево

uzhastik commented on ydb/core/fq/libs/row_dispatcher/topic_session.cpp:
```diff
@@ -556,6 +560,24 @@ void TTopicSession::CloseTopicSession() {
 void TTopicSession::TTopicEventProcessor::operator()(NYdb::NTopic::TReadSessionEvent::TDataReceivedEvent& event) {
+    auto minTime = Self.GetMinStartingMessageTimestamp();
+    auto &messages = event.GetMessages();
+    auto oldsize = messages.size();
```
локальные переменные именуем в camelCasing: oldSize

uzhastik commented on ydb/core/fq/libs/row_dispatcher/topic_session.cpp:
```diff
@@ -556,6 +560,24 @@ void TTopicSession::CloseTopicSession() {
+    if (!messages.empty()) {
+        Self.LastMessageOffset = messages.back().GetOffset();
+    }
+    messages.erase(std::remove_if(messages.begin(), messages.end(),
```
давай заложимся на монотонность GetWriteTime и сделаем оптимальнее. проверим один элемент для начала. если он больше minTime, то все сообщения нужны. если меньше, то бинарным поисков найдем линию отреза

yumkam replied:
Ошибка rebase (это уже решено в main)

uzhastik commented on ydb/core/fq/libs/row_dispatcher/topic_session.cpp:
```diff
@@ -556,6 +560,24 @@ void TTopicSession::CloseTopicSession() {
+        Self.LastMessageOffset = messages.back().GetOffset();
+    }
+    messages.erase(std::remove_if(messages.begin(), messages.end(),
+                [minTime](const auto &message) {
```
амперсанд прижимаем влево

uzhastik commented on ydb/core/fq/libs/row_dispatcher/ut/row_dispatcher_ut.cpp:
```diff
@@ -138,12 +139,14 @@ class TFixture : public NUnitTest::TBaseFixture {
             "Token",
             {},         // readOffset,
+            {{"","","",NYdb::NFederatedTopic::TFederatedTopicClient::TClusterInfo::EStatus::AVAILABLE}},
+            {*std::max_element(partitionIds.begin(), partitionIds.end()) + 1},
```
давай ассерт поставим на непустоту коллекции

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_rd_read_actor.cpp:
```diff
@@ -553,11 +622,12 @@ TDuration TDqPqRdReadActor::GetCpuTime() {
 
 std::vector<ui64> TDqPqRdReadActor::GetPartitionsToRead() const {
     std::vector<ui64> res;
+    ui64 partitionsCount = PartitionsCount.size() ? *std::max_element(PartitionsCount.cbegin(), PartitionsCount.cend()) : ReadParams.GetPartitioningParams().GetTopicPartitionsCount();
```
для проверки на пустоту контейнера есть спец метод empty()

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_rd_read_actor.cpp:
```diff
@@ -1003,7 +1079,7 @@ void TDqPqRdReadActor::UpdateSessions() {
         auto& session = Sessions.at(rowDispatcherActorId);
         SRC_LOG_I("Create session to " << rowDispatcherActorId << ", generation " << session.Generation);
         for (auto partitionId : partitions) {
```
давай цикл заменим на insert с итераторами

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_rd_read_actor.cpp:
```diff
@@ -1021,6 +1097,61 @@ void TDqPqRdReadActor::UpdateQueuedSize() {
+        auto describeTopicFutures = std::make_shared<std::vector<NYdb::NTopic::TAsyncDescribeTopicResult>>();
+        describeTopicFutures->reserve(federatedClusters.size());
+        
+        for (auto &info: federatedClusters) {
```
амперсанды прижимаем влево

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_rd_read_actor.cpp:
```diff
@@ -1021,6 +1097,61 @@ void TDqPqRdReadActor::UpdateQueuedSize() {
+        }
+        // FIXME don't WaitAll, receive all info desynced
+        NThreading::WaitAll(*describeTopicFutures).Subscribe(
+            [describeTopicFutures, actorSystem, selfId, federatedClusters = std::move(federatedClusters), LogPrefix = std::move(LogPrefix)](const auto& ) mutable {
```
зачем тебе тут logPrefix?

uzhastik replied:
и зачем делать std::move?

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_read_actor.cpp:
```diff
@@ -87,10 +91,30 @@ struct TEvPrivate {
     struct TEvSourceDataReady : public TEventLocal<TEvSourceDataReady, EvSourceDataReady> {};
     struct TEvReconnectSession : public TEventLocal<TEvReconnectSession, EvReconnectSession> {};
+    struct TEvReceivedClusters : public NActors::TEventLocal<TEvReceivedClusters, EvReceivedClusters> {
+        TEvReceivedClusters(
```
explicit

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_read_actor.cpp:
```diff
@@ -122,8 +146,20 @@ class TDqPqReadActor : public NActors::TActor<TDqPqReadActor>, public NYql::NDq:
     };
 
+    struct TClusterState {
+        ui32 Index;
```
старайся инициализировать скалярные поля, чтобы проблем под санитайзером не было

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_read_actor.cpp:
```diff
@@ -281,6 +348,86 @@ class TDqPqReadActor : public NActors::TActor<TDqPqReadActor>, public NYql::NDq:
     }
 
+
+    NThreading::TFuture<std::vector<NYdb::NFederatedTopic::TFederatedTopicClient::TClusterInfo>> AsyncInit;
```
поля класса давать держать в одной месте

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_read_actor.cpp:
```diff
@@ -281,6 +348,86 @@ class TDqPqReadActor : public NActors::TActor<TDqPqReadActor>, public NYql::NDq:
+    void Handle(TEvPrivate::TEvDescribeTopicResult::TPtr& ev) {
+        auto clusterIndex = ev->Get()->ClusterIndex;
+        auto partitionCount = ev->Get()->PartitionCount;
+        if(auto status = ev->Get()->Status) {
```
нужен пробел после if

uzhastik commented on ydb/library/yql/providers/pq/async_io/dq_pq_read_actor.cpp:
```diff
@@ -369,36 +527,43 @@ class TDqPqReadActor : public NActors::TActor<TDqPqReadActor>, public NYql::NDq:
     }
 
-    NYdb::NTopic::TReadSessionSettings GetReadSessionSettings() const {
+    NYdb::NTopic::TReadSessionSettings GetReadSessionSettings(TClusterState &clusterState) const {
```
амперсанд прижимаем влево

uzhastik commented on ydb/library/yql/providers/pq/async_io/ya.make:
```diff
@@ -25,6 +25,7 @@ PEERDIR(
     ydb/public/sdk/cpp/adapters/issue
     ydb/public/sdk/cpp/src/client/driver
     ydb/public/sdk/cpp/src/client/topic
+    ydb/public/sdk/cpp/src/client/federated_topic
```
sort plz

uzhastik commented on ydb/library/yql/providers/pq/common/ya.make:
```diff
@@ -3,6 +3,7 @@ LIBRARY()
 SRCS(
     pq_meta_fields.cpp
     yql_names.cpp
+    pq_partition_key.cpp
```
sort plz

uzhastik commented on ydb/library/yql/providers/pq/common/pq_partition_key.h:
```diff
@@ -0,0 +1,26 @@
+    struct TPartitionKey {
+        TString Cluster;
+        ui64 PartitionId;
+        bool operator==(const TPartitionKey& other) const {
```
= default ?

uzhastik commented on ydb/library/yql/providers/pq/common/pq_partition_key.h:
```diff
@@ -0,0 +1,26 @@
+}
+
+template <>
+inline void Out<NPq::TPartitionKey>(IOutputStream& stream, TTypeTraits<NPq::TPartitionKey>::TFuncParam& t) {
```
зачем тут TTypeTraits ? нельзя явно написать const TPartitionKey& t ?

yumkam replied:
Простые варианты ему почему-то не нравились

uzhastik commented on ydb/library/yql/providers/pq/gateway/dummy/yql_pq_dummy_gateway.cpp:
```diff
@@ -8,6 +8,23 @@
+        return NThreading::MakeFuture(std::move(dbInfo));
+    }
+
+    ~TDummyFederatedTopicClient() {}
```
бесполезное объявление

uzhastik commented on ydb/library/yql/providers/pq/gateway/native/yql_pq_gateway.cpp:
```diff
@@ -152,6 +154,27 @@ NYdb::NTopic::TTopicClientSettings TPqNativeGateway::GetTopicClientSettings() co
+
+NYdb::NFederatedTopic::TFederatedTopicClientSettings TPqNativeGateway::GetFederatedTopicClientSettings() const {
+    NYdb::NFederatedTopic::TFederatedTopicClientSettings settings;
+    if (CommonTopicClientSettings) {
```
выходи из функции пораньше, старайся не делать длинных вложенных скоупов
```
if (!CommonTopicClientSettings) {
  return {};
}
NYdb::NFederatedTopic::TFederatedTopicClientSettings settings;
settings.DefaultCompressionExecutor(CommonTopicClientSettings->DefaultCompressionExecutor_);
settings.DefaultHandlersExecutor(CommonTopicClientSettings->DefaultHandlersExecutor_);
....
return settings;
```

uzhastik commented on ydb/library/yql/providers/pq/gateway/native/yql_pq_session.cpp:
```diff
@@ -85,14 +98,45 @@ NPq::NConfigurationManager::TAsyncDescribePathResult TPqSession::DescribePath(co
+            return allFutureDescribes.Apply([results = std::move(results), paths = std::move(paths), allClustersInfo = std::move(allClustersInfo), cluster, database, path](const auto& ) mutable {
+                ui32 partitions = 0;
+                yexception ex;
+                TMutex mutex;
```
мьютекс я бы удалил

uzhastik commented on ydb/library/yql/providers/pq/gateway/native/yql_pq_session.cpp:
```diff
@@ -85,14 +98,45 @@ NPq::NConfigurationManager::TAsyncDescribePathResult TPqSession::DescribePath(co
+                        ex << "Failed to describe topic `" << cluster << "`.`" << path << "` (name '" << allClustersInfo[i].Name << "' endpoint '" << allClustersInfo[i].Endpoint << "' path '" << paths[i] << "') in the database `" << database << "`: " << describeTopicResult.GetIssues().ToString();
+                        continue;
+                    }
+                    partitions = std::max(partitions, describeTopicResult.GetTopicDescription().GetTotalPartitionsCount());
```
uint32_t и ui32 могут быть разными типам и тогда компиляция сломается. лучше гарантируй одинаковые типы или явно напиши std::max<ui32>()

uzhastik commented on ydb/library/yql/providers/pq/provider/yql_pq_topic_client.h:
```diff
@@ -31,6 +32,13 @@ class ITopicClient : public TThrRefBase {
+public:
+    using TPtr = TIntrusivePtr<IFederatedTopicClient>;
+
+    virtual NThreading::TFuture<std::vector<NYdb::NFederatedTopic::TFederatedTopicClient::TClusterInfo>> GetAllTopicClusters() = 0;
```
можно сделать const?

yumkam replied:
Нет, он делает запрос, а не получает уже имеющуюся информацию

yumkam committed:
spaces

yumkam committed:
fix thinko in sample functions

yumkam committed:
move AVAILABLE|READ_ONLY to function

yumkam committed:
Merge branch 'ydb-sdk-add-federated-topic-get-all-clusters' into logbroker-federation-hacks-v4

yumkam committed:
remove unnecessary std::move

yumkam committed:
fix potential circular shared_ptr dependency

yumkam committed:
Merge branch 'ydb-sdk-add-federated-topic-get-all-clusters' into logbroker-federation-hacks-v4

yumkam committed:
spaces

yumkam committed:
get rid of future.Apply([future]...

this is prone to circular dependency (and size of copied values is
expected to be small)

yumkam committed:
fix rebase glitch

different patch was applied upstream

yumkam committed:
resolve TODO

yumkam committed:
spacing

yumkam committed:
Add comment about GetAllClusters

yumkam committed:
Merge branch 'ydb-sdk-add-federated-topic-get-all-clusters' into logbroker-federation-hacks-v4

yumkam committed:
review issues

yumkam committed:
Set StartingMessageTimestamp at start of session

(cherry picked from commit c26a0296a3fe47806a7b71f658083c38df0996b8)

yumkam committed:
fixed rd to use gateway and get rid of NFuture::WaitAll

yumkam committed:
pq describe: skip unavailable clusters

yumkam committed:
Add comment about unsolved circular dependency

yumkam commented on ydb/library/yql/providers/pq/gateway/native/yql_pq_session.cpp:
```diff
@@ -85,14 +98,44 @@ NPq::NConfigurationManager::TAsyncDescribePathResult TPqSession::DescribePath(co
-            NPq::NConfigurationManager::TTopicDescription desc(path);
-            desc.PartitionsCount = describeTopicResult.GetTopicDescription().GetTotalPartitionsCount();
-            return NPq::NConfigurationManager::TDescribePathResult::Make<NPq::NConfigurationManager::TTopicDescription>(std::move(desc));
+            auto allFutureDescribes = NThreading::WaitAll(results);
```
Здесь все ещё есть циклическая зависимость:
До тех пор, пока фьюча не сработает:
- фьючи из results ссылаются на allFutureDescribe (которая должна сработать после них)
- lamba содержит results и ссылается на них
- allFutureDescribe содержит lambda
Как только фьюча выстреливает, цепочка зависимостей распадается

kardymonds commented on ydb/core/fq/libs/row_dispatcher/row_dispatcher.cpp:
```diff
@@ -139,24 +140,26 @@ ui64 MaxSessionBufferSizeBytes = 16000000;
         TString Database;
         TString TopicPath;
-        ui64 PartitionId;
+        TPartitionKey PartitionKey;
```
У кластеров Endpoint же разные? Если да, то зачем тут использовать Cluster (в PartitionKey). Т.е. как будто Cluster не требуется в TRowDispatcher/TTopicSession .

yumkam replied:
dq_pq_rd_read_actor получает пару кластер + номер партиции с message batch;

(fwiw, я в процессе очередного переигрывания :-|, с уборкой cluster awareness из row_dispatcher и замены TDqRdReadActor на коллекцию акторов, но когда оно сойдётся и насколько будет приемлемо пока не понятно)

kardymonds commented on ydb/tests/fq/pq_async_io/ut/dq_pq_rd_read_actor_ut.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <ydb/library/yql/dq/common/rope_over_buffer.h>
 #include <library/cpp/testing/unittest/gtest.h>
 #include <library/cpp/testing/unittest/registar.h>
+#include <ydb/tests/fq/pq_async_io/mock_pq_gateway.h>
```
А нет рецепта с федеративным логброкером чтобы интеграционный теста написать?

yumkam replied:
Я пока не очень понял, а как эти тесты вообще работают; как-то запинал чтобы компилировались и не падали, но в каком месте вот этот mock_pq_gateway используется пока не понял :-|

kardymonds replied:
Сейчас  для старого read actor (в main) эти тесты работают с реальным ydb_recipe (в ya,make указан).
Для rd read actor они просто эвенты проверяют.
Для topic_session тесты также с реальным ydb_recipe запукаются. И только последний тест для topic_session написан с замоканным гейтвеем (RestartSessionIfQueryStopped, но тест вроде нестабильны)

yumkam replied:
Насколько я понимаю, федерированый логброкер в реальном ydb симулировать не получиться, а вот намокать что-то можно?

yumkam committed:
DescribePath: add dummy placeholder for offline cluster

yumkam committed:
cosmetics

yumkam committed:
cosmetics

yumkam committed:
cosmetics

yumkam committed:
improve error handling

yumkam committed:
improve error handling (DescribePath)

yumkam committed:
improve error handling (TDqPqRdReadActor)

yumkam committed:
Merge remote-tracking branch 'origin/main' into logbroker-federation-hacks-v4

Conflicts:
	ydb/public/sdk/cpp/include/ydb-cpp-sdk/client/federated_topic/federated_topic.h
	ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic.cpp
	ydb/public/sdk/cpp/src/client/federated_topic/impl/federated_topic_impl.cpp
